dbus_next/_private/unmarshaller.py

```diff
@@ -1,307 +1,348 @@
+
+        if endian != LITTLE_ENDIAN and endian != BIG_ENDIAN:
+            raise InvalidMessageError(
+                f"Expecting endianness as the first byte, got {endian} from {buffer}"
+            )
         if protocol_version != PROTOCOL_VERSION:
-            raise InvalidMessageError(f'got unknown protocol version: {protocol_version}')
-
-        body_len = self.read_uint32()
-        serial = self.read_uint32()
-
-        header_len = self.read_uint32()
-        msg_len = header_len + self._padding(header_len, 8) + body_len
-        self.read(msg_len, prefetch=True)
-        # backtrack offset since header array length needs to be read again
-        self.offset -= 4
-
-        header_fields = {}
-        for field_struct in self.read_argument(SignatureTree._get('a(yv)').types[0]):
-            field = HeaderField(field_struct[0])
-            header_fields[field.name] = field_struct[1].value
-
-        self.align(8)
-
-        path = header_fields.get(HeaderField.PATH.name)
-        interface = header_fields.get(HeaderField.INTERFACE.name)
-        member = header_fields.get(HeaderField.MEMBER.name)
-        error_name = header_fields.get(HeaderField.ERROR_NAME.name)
-        reply_serial = header_fields.get(HeaderField.REPLY_SERIAL.name)
-        destination = header_fields.get(HeaderField.DESTINATION.name)
-        sender = header_fields.get(HeaderField.SENDER.name)
-        signature = header_fields.get(HeaderField.SIGNATURE.name, '')
-        signature_tree = SignatureTree._get(signature)
-        # unix_fds = header_fields.get(HeaderField.UNIX_FDS.name, 0)
-
-        body = []
-
-        if body_len:
-            for type_ in signature_tree.types:
-                body.append(self.read_argument(type_))
-
-        self.message = Message(destination=destination,
-                               path=path,
-                               interface=interface,
-                               member=member,
-                               message_type=message_type,
-                               flags=flags,
-                               error_name=error_name,
-                               reply_serial=reply_serial,
-                               sender=sender,
-                               unix_fds=self.unix_fds,
-                               signature=signature_tree,
-                               body=body,
-                               serial=serial)
+            raise InvalidMessageError(
+                f"got unknown protocol version: {protocol_version}"
+            )
+
+        self.body_len, self.serial, self.header_len = UNPACK_LENGTHS[
+            endian
+        ].unpack_from(buffer, 4)
+        self.msg_len = (
+            self.header_len + (-self.header_len & 7) + self.body_len
+        )  # align 8
+        if (sys.byteorder == "little" and endian == LITTLE_ENDIAN) or (
+            sys.byteorder == "big" and endian == BIG_ENDIAN
+        ):
+            self.can_cast = True
```

Comment:
Instead of can cast remap read argument since it's the only place used. This should reduce some more branching